src/Features/CSharp/Portable/Snippets/CSharpForLoopSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+                if (variableDeclarator != null)
+                {
+                    var declaratorIdentifier = variableDeclarator.Identifier;
+                    placeHolderBuilder.Add(declaratorIdentifier.ToString(), new List<int>() { declaratorIdentifier.SpanStart });
+                }
+            }
+
+            // Gets the placeholders present in the left and right side of the conditional.
+            if (condition != null)
+            {
+                var conditionExpression = (BinaryExpressionSyntax)condition;
+                var left = conditionExpression.Left;
+                if (!placeHolderBuilder.TryGetValue(left.ToString(), out var incrementorList))
+                {
+                    incrementorList = new List<int>();
+                    placeHolderBuilder.Add(left.ToString(), incrementorList);
+                }
+
+                incrementorList.Add(left.SpanStart);
+
+                var right = conditionExpression.Right;
+                if (!placeHolderBuilder.TryGetValue(right.ToString(), out var lengthList))
+                {
+                    lengthList = new List<int>();
+                    placeHolderBuilder.Add(right.ToString(), lengthList);
+                }
+
+                lengthList.Add(right.SpanStart);
+            }
+
+            // Gets the placeholder present in the incrementor.
+            if (incrementor != null)
+            {
+                var operand = ((PostfixUnaryExpressionSyntax)incrementor).Operand;
+                if (!placeHolderBuilder.TryGetValue(operand.ToString(), out var incrementorList))
+                {
+                    incrementorList = new List<int>();
+                    placeHolderBuilder.Add(operand.ToString(), incrementorList);
+                }
+
+                incrementorList.Add(operand.SpanStart);
+            }
+
+            foreach (var kvp in placeHolderBuilder)
+            {
+                arrayBuilder.Add(new SnippetPlaceholder(kvp.Key, kvp.Value.ToImmutableArray()));
+            }
+
+            return arrayBuilder.ToImmutableArray();
+
+        }
+
+        /// <summary>
+        /// Gets the start of the BlockSyntax of the for statement
+        /// to be able to insert the caret position at that location.
+        /// </summary>
+        protected override int GetCaretPosition(SyntaxNode caretTarget)
+        {
+            GetPartsOfForStatement(caretTarget, out _, out _, out _, out var statement);
+            return statement.SpanStart + 1;
+
+        }
+
+        private static void GetPartsOfForStatement(SyntaxNode node, out SyntaxNode? declaration, out SyntaxNode? condition,
+            out SyntaxNode? incrementor, out SyntaxNode statement)
+        {
+            var forStatement = (ForStatementSyntax)node;
+            declaration = forStatement.Declaration;
+            condition = forStatement.Condition;
+            // We can assume there will only be one incrementor since it is only constructed with one.
```

Comment:
if you can assume this, why can't you assume things like that the Declaration/Condition will be non-null?